lisa/trace.py

```diff
@@ -294,7 +314,812 @@ def get_view(self, window):
+                    path = os.path.join(swap_dir, filename)
+                    try:
+                        swap_entry = PandasDataSwapEntry.from_path(path)
+                    # If there is any issue with that entry, just ignore it
+                    except Exception:
+                        continue
+                    else:
+                        yield (swap_entry.pd_desc_nf, swap_entry)
+
+            swap_content = dict(load_swap_content(swap_dir))
+
+            metadata_ = mapping['metadata']
+            metadata = {**metadata_, **metadata}
+
+        return cls(swap_content=swap_content, swap_dir=swap_dir, metadata=metadata, trace_path=trace_path, trace_md5=new_md5, **kwargs)
+
+    def to_swap_dir(self):
+        """
+        Write the persistent state to the swap area if any, no-op otherwise.
+        """
+        if self.swap_dir:
+            path = os.path.join(self.swap_dir, self.TRACE_META_FILENAME)
+            self.to_path(path)
+
+    @classmethod
+    def from_swap_dir(cls, swap_dir, **kwargs):
+        """
+        Reload the persistent state from the given ``swap_dir``.
+
+        :Variable keyword arguments: Forwarded to :class:`TraceCache`.
+        """
+        if swap_dir:
+            try:
+                return cls._from_swap_dir(swap_dir=swap_dir, **kwargs)
+            except (FileNotFoundError, TraceCacheSwapVersionError, json.decoder.JSONDecodeError):
+                pass
+
+        return cls(swap_dir=swap_dir, **kwargs)
+
+    def _estimate_data_swap_cost(self, data):
+        return self._estimate_data_swap_size(data) * self.swap_cost
+
+    def _estimate_data_swap_size(self, data):
+        return self._data_mem_usage(data) * self._data_mem_swap_ratio
+
+    def _update_data_swap_size_estimation(self, data, size):
+        old_ratio = self._data_mem_swap_ratio
+        mem_usage = self._data_mem_usage(data)
+        if mem_usage:
+            new_ratio = size / mem_usage
+            alpha = 0.75
+            self._data_mem_swap_ratio = (1 - alpha) * old_ratio + alpha * new_ratio
+
+    def _data_mem_usage(self, data):
+        mem = data.memory_usage()
+        try:
+            return mem.sum()
+        except AttributeError:
+            return mem
+
+    def _should_evict_to_swap(self, pd_desc, data):
+        compute_cost = self._data_cost.get(pd_desc, 0)
+        swap_cost = self._estimate_data_swap_cost(data)
+        return swap_cost <= compute_cost
+
+    def _swap_path_of(self, pd_desc):
+        if self.swap_dir:
+            pd_desc_nf = pd_desc.normal_form
+            swap_entry = self._swap_content[pd_desc_nf]
+            filename = swap_entry.data_filename
+            return os.path.join(self.swap_dir, filename)
+        else:
+            raise ValueError('Swap dir is not setup')
+
+    def _update_swap_cost(self, new_cost):
+        old_cost = self.swap_cost
+        # EWMA to keep a relatively stable cost
+        if old_cost != self.INIT_SWAP_COST:
+            alpha = 0.75
```

Comment:
> Is that roughly what we want ?

I don't really know. What I definitely don't want is to replace the cost with the new value, as it could be biased by a better compression ratio for a given dataframe or stuff like that. I also want the initial arbitrary cost to fade over time. Ultimately, I don't expect that value to be accurate, but I expect it to be of the right order of magnitude.

I could factor that out in a small helper indeed.